src/evaluators/BinaryExpression.js

```diff
@@ -190,17 +173,48 @@ export function computeBinary(
   }
 
   if (lval instanceof AbstractValue || rval instanceof AbstractValue) {
-    // generate error if binary operation might throw or have side effects
     if (realm.isInPureScope()) {
-      realm.evaluateWithPossibleThrowCompletion(
-        () => {
-          getPureBinaryOperationResultType(realm, op, lval, rval, lloc, rloc);
-          return realm.intrinsics.undefined;
-        },
-        TypesDomain.topVal,
-        ValuesDomain.topVal
-      );
+      // If we're in pure mode we can recover even if this operation might not be pure.
+      // To do that, we'll temporarily override the error handler.
+      const previousErrorHandler = realm.errorHandler;
+      let isPure = true;
+      realm.errorHandler = diagnostic => {
+        isPure = false;
+        return "Recover";
+      };
+      try {
+        const valueType = getPureBinaryOperationResultType(realm, op, lval, rval, lloc, rloc);
+        if (!isPure) {
+          // If this ended up reporting an error, it might not be pure, so we'll leave it in
+          // as a temporal operation with a known return type.
+
+          // Some of these values may trigger side-effectful user code such as valueOf.
+          // To be safe, we have to Havoc them. However, "in" and "instanceof" cannot
+          // have side-effects other than throw so we don't need to havoc them.
+          if (op !== "in" && op !== "instanceof") {
+            Havoc.value(realm, lval, loc);
+            Havoc.value(realm, rval, loc);
+          }
+          return realm.evaluateWithPossibleThrowCompletion(
+            () =>
+              AbstractValue.createTemporalFromBuildFunction(
+                realm,
+                valueType,
+                [lval, rval],
+                ([lnode, rnode]: Array<BabelNodeExpression>) => {
+                  // TODO: How can I communicate the loc of the original nodes for source maps.
```

Comment:
You may as well pass it in as an optional parameter like we do for the two operands.

Reply:
We don't have a way to do this today though? I don't see anywhere this is done for abstract expressions. Seems like that'll have to be a larger issue to go back and provide source map hints for all these residual abstract expressions.